File: algoexpert/src/stacks/medium/NextGreaterElement.java
package stacks.medium;

import java.util.*;

public class NextGreaterElement {
  public int[] nextGreaterElement(int[] array) {
    Stack<Integer> st = new Stack<>();
    int[] res = new int[array.length];
    Arrays.fill(res, -1);

    for(int i = array.length - 1; i >= 0; i--) {
      while(!st.isEmpty() && st.peek() <= array[i]) st.pop();
      res[i] = st.isEmpty() ?  -1 : st.peek();
      st.add(array[i]);
    }

    for(int i = array.length - 1; i >= 0; i--) {
      while(!st.isEmpty() && st.peek() <= array[i]) st.pop();
      res[i] = st.isEmpty() ?  -1 : st.peek();
      st.add(array[i]);
    }

    return res;
  }
}

File: algoexpert/src/stacks/medium/next_greater_element.ts
export function nextGreaterElement(array: number[]) {
  const stack: number[] = [];
  const result: number[] = new Array(array.length);

  const loopArray = () => {
    for(let i = array.length - 1; i >= 0; i--) {
      let nextMajor = -1;
      while(stack.length > 0) {
        if (stack[stack.length -1] <= array[i]) stack.pop();
        else {
          nextMajor = stack[stack.length - 1];
          break;
        }
      }

      stack.push(array[i]);
      result[i] = nextMajor;
    }
  }

  loopArray();
  loopArray();
  return result;
}

